### PR TITLE
Add `tar_stars()`

### PR DESCRIPTION
**Pull request by brownag**
Adds format methods `tar_stars()`, `tar_stars_proxy()` for #14 

 - Adds {sf}, {stars} and {ncmeta} to suggests
 - When `mdim=TRUE` `tar_stars()` and `tar_stars_proxy()` can use the GDAL Multidimensional Raster Data Model as a backend via `stars::read/write_mdim()`instead of `stars::read/write_stars()`
 - When `ncdf=TRUE` "read" methods can be toggled to use `stars::read_ncdf()` (instead of `stars::read_stars()` or `stars::read_mdim()`) which requires {ncmeta} 

 - To be less repetitive I think`tar_stars_proxy()` could call `tar_stars(proxy=TRUE)` but I still need to sort out how arguments would need to be deparsed to do this.

### CODE REVIEW

File: R/tar-stars.R
Comment (Aariq) [inline R/tar-stars.R:98]: `sf` is in Depends for `stars`, so this is probably not necessary

File: R/tar-stars.R
Comment (Aariq) [inline R/tar-stars.R:102]: Maybe move this to the very top of `tar_stars()`

File: R/tar-stars.R
Comment (Aariq) [inline R/tar-stars.R:122]: Can't `stars` objects contain vectors?  How does this work with vector data cubes?

File: tests/testthat/test-tar-stars.R
Comment (Aariq) [inline tests/testthat/test-tar-stars.R:3]: I think this is unecessary, again because `sf` is a hard dependency of `stars`

**Review by Aariq — COMMENTED** [review]
It might be good to clarify that this currently works for stars raster cubes, but might not work for vector cubes or stars.proxy objects. We might want to explore these other formats a bit and depending on the findings of that exploration, perhaps call this `tar_stars_raster` even though both raster cubes and vector cubes have the same class (`stars`).

File: R/tar-stars.R
Comment (brownag) [reply R/tar-stars.R:98]: You are right, it was a reflex for me b/c I was using `sf::st_drivers()`, thanks for pointing this out, will fix

File: R/tar-stars.R
Comment (brownag) [reply R/tar-stars.R:102]: will do to mirror your other changes/suggestions!

File: R/tar-stars.R
Comment (njtierney) [inline R/tar-stars.R:10]: We don't need to use `#' @importFrom rlang <fun>` each time we want to use those functions - I suggest that we move these to https://github.com/njtierney/geotargets/blob/master/R/geotargets-package.R#L4-L5

File: R/tar-stars.R
Comment (njtierney) [inline R/tar-stars.R:102]:
```suggestion
check_installed("stars")
```

**Review by njtierney — CHANGES_REQUESTED** [review]
some suggestions on pkg documentation and using `rlang::check_installed()`

File: R/tar-stars.R
Comment (brownag) [reply R/tar-stars.R]: done in https://github.com/njtierney/geotargets/pull/33/commits/0e20c172f574efed855625ec06219c40eade9917

File: R/tar-stars.R
Comment (brownag) [reply R/tar-stars.R]: done in https://github.com/njtierney/geotargets/pull/33/commits/0e20c172f574efed855625ec06219c40eade9917 and https://github.com/njtierney/geotargets/pull/33/commits/78bdee5347f99818837669bac4aa75e69911f6f4

File: tests/testthat/test-tar-stars.R
Comment (brownag) [reply tests/testthat/test-tar-stars.R]: done in https://github.com/njtierney/geotargets/pull/33/commits/0e20c172f574efed855625ec06219c40eade9917

File: R/tar-stars.R
Comment (brownag) [reply R/tar-stars.R]: done in https://github.com/njtierney/geotargets/pull/33/commits/0e20c172f574efed855625ec06219c40eade9917

File: R/tar-stars.R
Comment (brownag) [reply R/tar-stars.R]: done in https://github.com/njtierney/geotargets/pull/33/commits/0e20c172f574efed855625ec06219c40eade9917

**Comment by brownag** [comment]
> It might be good to clarify that this currently works for stars raster cubes, but might not work for vector cubes or stars.proxy objects. 

Good point about vector cubes, we should work up an example for that in a new issue. We can hash it out before merging this PR if need be.

In https://github.com/njtierney/geotargets/pull/33/commits/0e20c172f574efed855625ec06219c40eade9917 I added support for `stars_proxy` objects. `tar_stars()` now takes a `proxy` argument (default: FALSE) and there is a `tar_stars_proxy()` function for shorthand.  I _think_ `tar_stars_proxy()` could call `tar_stars()` but would need to think more about the deparsing/evaluation of the arguments to make it work.

> We might want to explore these other formats a bit and depending on the findings of that exploration, perhaps call this `tar_stars_raster` even though both raster cubes and vector cubes have the same class (`stars`).

I think we should open an issue for this and look into some vector data cube examples--I am not a heavy user of stars, and can't say I a ton of experience with that specific application--though I want to investigate now. 

As I understand it creating vector data cubes requires reading the vector data with `sf::st_read()`/`sf::read_sf()` and then converting to stars with `stars::st_as_stars()`. So, we definitely might want to provide a target that eases this process, but right now if the user has a stars or stars_proxy object already made, I _think_ what is implemented in this PR "works"--open to a counterexample!

File: R/tar-stars.R
Comment (brownag) [reply R/tar-stars.R:262]: Yes, good point, `stars` objects can contain vector geometry and an arbitrary number of dimensions. I said "new" issue but we can discuss this approach more in #14, I think

File: R/tar-stars.R
Comment (Aariq) [inline R/tar-stars.R:8]: Just noticing that sometimes this arg is called `options` and sometimes it is called `gdal`.  We should decide whether we want to be consistent throughout geotargets or if we want to be consistent with whatever function the arg is being passed to (or split the difference and call it `gdal_options`)

Comment (brownag) [reply R/tar-stars.R:6]: @Aariq 

> Just noticing that sometimes this arg is called `options` and sometimes it is called `gdal`. We should decide whether we want to be consistent throughout geotargets or if we want to be consistent with whatever function the arg is being passed to (or split the difference and call it `gdal_options`)

This is a good idea. I was sortof leaning towards emulating whatever the "write" function uses for each particular spatial backend being used... But that won't cover everything. I definitely see the value of choosing something consistent across geotargets.

Comment (brownag) [reply R/tar-stars.R:6]: Resolving this, lets discuss in #44

**Comment by Aariq** [comment]
If you'd like, I can attempt to bring this up to speed with the master branch.  That'll make it easier to review and discuss.

**Comment by brownag** [comment]
> If you'd like, I can attempt to bring this up to speed with the master branch. That'll make it easier to review and discuss.

Thanks for the offer, I've been trying to keep up with changes... and wanted to make sure I went over everything so got the new master branch merged in here.

I am going to re-request review on this now. I think all prior comments/changes have been resolved and/or referred to in an open issue.

File: R/tar-stars.R
Comment (Aariq) [inline R/tar-stars.R:7]:
```suggestion
#' @param proxy logical. Passed to [stars::read_stars()]. If `TRUE` the target will be read as an object of class `stars_proxy`. Otherwise, the object is class `stars`.
```

Seems like there is a missing word here maybe?

File: R/tar-stars.R
Comment (Aariq) [inline R/tar-stars.R:211]: Rather than `eval(substitute(...` we now have the option to pass env vars to the custom format using `targets::tar_resources_custom_format()`.  See the updates to `tar_terra_rast()`, e.g. [here](https://github.com/njtierney/geotargets/blob/e1b039f372f5f82c9cd08ac27c319c08955f0c75/R/tar-terra-rast.R#L85-L91) and [here](https://github.com/njtierney/geotargets/blob/e1b039f372f5f82c9cd08ac27c319c08955f0c75/R/tar-terra-rast.R#L104-L108).  I think we should stick to this new pattern as it's a bit easier to comprehend, IMO.

**Review by Aariq — CHANGES_REQUESTED** [review]
Could you re-run devtools::document() to update tar_stars.Rd

File: R/tar-stars.R
Comment (brownag) [reply R/tar-stars.R:211]: > Rather than `eval(substitute(...` we now have the option to pass env vars to the custom format using `targets::tar_resources_custom_format()`. 

I have this mostly implemented, it does make the code look cleaner. 

I'm running into some issues with arguments not being properly passed when using `tar_resources_custom_format()` though. I tinkered with it a bit last night and will take a fresh look tonight. If I can't figure it out I will push the changes anyway to see if you can spot what I am doing wrong.

File: R/geotargets-package.R
Comment (njtierney) [inline R/geotargets-package.R:10]: importing `arg_match0`, `is_installed` and `check_installed` with `@importFrom` is only needed if we want to use those functions without namespacing them. My preference is that we don't @importFrom these functions as they aren't used very often.

It is necessary to importFrom `%||%` at the moment, although I note that in R 4.4.0 `%||%` is implemented natively, which is very cool.

File: tests/testthat/test-tar-stars.R
Comment (njtierney) [inline tests/testthat/test-tar-stars.R:1]: Could you add one of these uses of tar_tars into the README.Rmd?

**Review by njtierney — CHANGES_REQUESTED** [review]
Pending the changes from @Aariq and a minor change that I have about using `@importFrom` and a request to add an example of stars into the README, I think this is looking good :)

File: tests/testthat/test-tar-stars.R
Comment (brownag) [reply tests/testthat/test-tar-stars.R:1]: I have now updated NAMESPACE, added a tar_stars() example and did a several other general updates to README

**Review by njtierney — APPROVED** [review]

**Comment by njtierney** [comment]
Fantastic, thanks @brownag ! :)

File: R/tar-stars.R
Comment (brownag) [reply R/tar-stars.R]: It appears the source of my problem is that environment variables that are created via `tar_resources_custom_format()` are not available / not working properly in the `tar_format()` `read` method. 

I managed to localize the problems I was having to that, so for time being continuing to use `eval(substitute(...))` for read method dynamics. Using the env vars for write method, though. 

I have not yet had a chance to look into whether this is "expected" or if it is an issue in targets.

All tests are passing with expected values for the current code. If code to access env vars in read method is uncommented, the `tar_stars_proxy()` tests fail because default `proxy=FALSE` is used because the env var GEOTARGETS_PROXY is apparently unset.

File: R/geotargets-package.R
Comment (brownag) [reply R/geotargets-package.R]: I have removed all three from importFrom as requested.. Indeed, only using `is_installed()` for `@examplesIf` block to check for stars, which may not really be necessary but I think we don't want to assume suggested packages are installed, in general. `check_installed()` is not used but I initially thought we would use it, thanks for pointing that out. `arg_match0` is used in a couple places now for terra and stars drivers checks.

For imports I generally prefer to handle functions explicitly, as well as use double colon to access them; the latter because it is easier for humans to see what is coming from where, and the former to explicitly declare what is used by the package in the NAMESPACE--no matter how commonly it is used.

Comment (njtierney) [reply R/geotargets-package.R]: Thank you! :)

> For imports I generally prefer to handle functions explicitly, as well as use double colon to access them; the latter because it is easier for humans to see what is coming from where, and the former to explicitly declare what is used by the package in the NAMESPACE--no matter how commonly it is used.

I hear you! I prefer to use `pkg::fun()`, with the only exception being for imported infix operators, e.g., `%||%`. Another compelling case I've seen is for packages that make heavy use of ggplot2 functions or rlang metaprogramming, then you can use `@import ggplot2` etc.

 > explicitly declare what is used by the package in the NAMESPACE--no matter how commonly it is used.

I think this is a good idea, and it can be helpful to identify which functions are actually used from a package. It does have the potential to get a bit gnarly if you are making heavy use of importing from various packages though. I haven't seen much more development of the [itdepends](https://github.com/r-lib/itdepends) pkg, but I think tools like that would be really useful. The reason I don't use @importFrom very often is because of reasons stated in the [rpkgs book section on dependencies](https://r-pkgs.org/dependencies-in-practice.html#sec-dependencies-in-imports-r-code):

> @importFrom aaapkg aaa_fun. This does suppress the NOTE, but it also does more. This causes aaapkg to be loaded whenever your package is loaded. In contrast, if you use the approach we recommend, the aaapkg will only be loaded if your user does something that actually requires it. This rarely matters in practice, but it’s always nice to minimize or delay the loading of additional packages.

Anyway, that is a bit of an aside! I want to say thanks again for all your continued help with this project. I'm now curious if the `itdepends` project can be revived 🤔

File: R/tar-stars.R
Comment (Aariq) [inline R/tar-stars.R:278]: Use user-specified `resources` as in #67 
```suggestion
        resources = utils::modifyList(targets::tar_resources(
            custom_format = targets::tar_resources_custom_format(
                #these envvars are used in read and write functions of format
                envvars = c("GEOTARGETS_GDAL_RASTER_DRIVER" = driver,
                            "GEOTARGETS_GDAL_RASTER_CREATION_OPTIONS" =
                                paste0(options, collapse = ";"),
                            "GEOTARGETS_GDAL_RASTER_MDIM" = mdim,
                            "GEOTARGETS_PROXY" = proxy,
                            "GEOTARGETS_USE_NCMETA" = ncdf)
            )
        ), resources)
```

**Review by Aariq — CHANGES_REQUESTED** [review]
Looking good.  Only other thing is can you update NEWS.md now that we have it?

**Comment by brownag** [comment]
> Looking good. Only other thing is can you update NEWS.md now that we have it?

Thanks, done!

File: R/tar-stars.R
Comment (Aariq) [reply R/tar-stars.R]: I see that you've commented this well, so I'm going to go ahead and merge this PR.  I think once we have all had some more practice with metaprogramming with targets, we'll go through and harmonize things across functions a bit.

**Pull request merged**
